### PR TITLE
pollAndProduce: Wait for all futures at once

### PR DESCRIPTION
**Pull request by birdayz**
So far, in a loop, each future is produced and waited for. This is very
slow, especially in cases for large fan-outs. Now, all records are
produced first, before waiting for them.

See #297

- I changed produceMessage to accept multiple messages instead of a single message.
- I changed produceMessage to not call get() anymore, instead it returns the future. It is now up to the caller (=pollAndProduceMany) when to call get().
- pollAndProduceMany calls get() on all futures after they were submitted to the producer.

### Checklist

- [x] Documentation (if applicable) (i think not)
- [X] Changelog

### CODE REVIEW

**Review by astubbs — CHANGES_REQUESTED** [review]
Looks great! But can you give me write access to your branch, and I'll push for formatting that I've got setup.

I need to publish the formatting config so we're all aligned. Should setup https://editorconfig.org - haven't needed to so far :)

Or - can you do another PR to setup https://editorconfig.org and let's get that right first, then we can do this PR on top of it. Need to get it done some time anyway. https://github.com/confluentinc/parallel-consumer/issues/301

What IDE format settings are you using?

Ok, I just pushed https://github.com/confluentinc/parallel-consumer/pull/302, can you merge on that, and try formatting?

**Comment by astubbs** [comment]
Blocked by:
- https://github.com/confluentinc/parallel-consumer/pull/302

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:141]:
```suggestion
            for (ProducerRecord<K, V> rec: outMsgs) {
```
or similar - `record` is a restricted identifier

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:59]: several extra whitespace changes

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:143]: prefer to break out for readability and line numbers while debugging
```suggestion
		var future = producer.send(rec, callback);
                futures.add(ParallelConsumer.Tuple.pairOf(record, future));
```

File: CHANGELOG.adoc
Comment (astubbs) [inline CHANGELOG.adoc:26]:
```suggestion
* #298: Improve PollAndProduce performance by first producing all records, and then waiting for the produce results. Previously, this was done for each ProduceRecord individually.
```

**Review by astubbs — CHANGES_REQUESTED** [review]

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:141]: (check out the solar lint plugin for IDEA)

Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:141]: === Recommended IDEA Plugins

* AsciiDoc
* CheckStyle
* CodeGlance
* Rainbow Brackets
* SonarLint
* Lombok

**Comment by birdayz** [comment]
I hope i addressed all issues. Can you check?
Shall i squash?

**Comment by birdayz** [comment]
> Looks great! But can you give me write access to your branch, and I'll push for formatting that I've got setup.
> 
> I need to publish the formatting config so we're all aligned. Should setup https://editorconfig.org - haven't needed to so far :)
> 
> Or - can you do another PR to setup https://editorconfig.org and let's get that right first, then we can do this PR on top of it. Need to get it done some time anyway. #301
> 
> What IDE format settings are you using?
> 
> Ok, I just pushed #302, can you merge on that, and try formatting?

I merged that branch and re-formatted the two files with the IntelliJ editorconfig plugin.

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java
Comment (birdayz) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java]: done

File: CHANGELOG.adoc
Comment (birdayz) [reply CHANGELOG.adoc]: done, and moved it to 0.5.0.1

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java
Comment (birdayz) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java]: fixed

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java
Comment (birdayz) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java]: re-formatted it

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:25]: hmm, .editorconfig setup must have an issue, as it's introduction extra whitespace changes here. Something must not be quite sync'd with the default IDEA formatting.

**Comment by astubbs** [comment]
> Shall i squash?
You can if you want, otherwise I'll just do it on PR acceptance in GH.

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:25]:
```suggestion
public class ParallelEoSStreamProcessor<K, V> extends AbstractParallelEoSStreamProcessor<K, V> 
        implements ParallelStreamProcessor<K, V> {
```

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:53]:
```suggestion
    public void pollAndProduceMany(Function<PollContext<K, V>, List<ProducerRecord<K, V>>> userFunction, 
        Consumer<ConsumeProduceResult<K, V, K, V>> callback) {
```
format with the line break, see if it lines up again, or is removed?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:79]:
```suggestion
                throw new InternalRuntimeError("Error while waiting for produce results", e);
```

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:108]:
```suggestion
    public void pollAndProduce(Function<PollContext<K, V>, ProducerRecord<K, V>> userFunction, 
    Consumer<ConsumeProduceResult<K, V, K, V>> callback) {
```
and run formatter, as above

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:83]:
```suggestion
                throw new IllegalArgumentException("Using non-transactional producer option, but Producer has a transaction ID - " + "
                the Producer must not have a transaction ID for this option. This is because having such an ID forces the " + 
                "Producer into transactional mode - i.e. you cannot use it without using transactions.");
```

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:118]: was this you, or the formatter that moved the <p>? Rest of the project has <p> on the blank line

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:307]: same question here with the formatting

Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:307]:
```suggestion
        if (!producerTransactionLock.isWriteLockedByCurrentThread()) 
            throw new IllegalStateException("Not held be me");
```

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:74]:
```suggestion
                    var recordMetadata = TimeUtils.time(() ->
                        future.getRight().get(options.getSendTimeout().toMillis(), TimeUnit.MILLISECONDS));
```

**Review by astubbs — CHANGES_REQUESTED** [review]

**Comment by birdayz** [comment]
i tried to get rid of the diffs manually. re-formatting in intelli J 2022.1, even with the editorconfig, does not seem to produce the same result. it would be nice to have a CLI command to run before committing, which formats.
from my PoV it can be merged, maybe you can run the formatter afterwards :)

**Comment by astubbs** [comment]
Thanks for the submission! 
Yes, it's not great. I'll look into what options there are going forward :) Thanks for persevering!

**Review by astubbs — APPROVED** [review]

**Comment by astubbs** [comment]
Released: https://github.com/confluentinc/parallel-consumer/releases/tag/0.5.1.0

**Pull request merged**
